### PR TITLE
feat(SEC-1293): Independent releases

### PR DESCRIPTION
**Pull request by pankajmouriyakong**
<Explain your change!>

Did you sign your commit? [Instructions](https://github.com/kumahq/.github/blob/main/CONTRIBUTING.md#sign-your-commits)

Have you read [Contributing guidelines](https://github.com/kumahq/.github/blob/main/CONTRIBUTING.md)?

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        shell: bash
+        run: pnpm i
+
+      - name: Release new versions
+        shell: bash
+        run: |
+          git checkout ${{ github.head_ref }}
+          git config --local user.name "${{ vars.GH_APP__KONG__BOT_USERNAME }}"
+          git config --local user.email "${{ vars.GH_APP__KONG__BOT_EMAIL }}"
+          pnpm version:dry-run
```
Comment (saisatishkarra) [inline .github/workflows/ci.yml:49]: This workflow must be changed to have a separate job for each package to  trigger based on `paths-changed-filter` GH action for each package and run the npm build step to ensure no node build errors rather than just installing dependencies and doing nothing. 

Ex: https://github.com/Kong/shared-actions/blob/main/.github/workflows/ci.yaml#L108-L173

Comment (saisatishkarra) [reply .github/workflows/ci.yml:49]: To limit the size of this CI build file over time as more independent job-per-package are added, you can use a pattern to use `workflow_call` for each package to build in a separate yml (reusable workflow) file and then use the workflow file within this build job. Helps with manageability of file size and keeps things decoupled better.

Ex: 
- Parent build workflow for mesh: https://github.com/Kong/kong-mesh/blob/master/.github/workflows/kuma-build-test-distribute.yaml
- Invokes reusable workflows to test: https://github.com/Kong/kong-mesh/blob/master/.github/workflows/kuma-build-test-distribute.yaml#L126 managed within a separate file https://github.com/Kong/kong-mesh/blob/master/.github/workflows/kuma-_test.yaml

Comment (saisatishkarra) [reply .github/workflows/ci.yml:49]: This is missing the `commitlint` to check the enforced commit structure before geenerating the changelog
https://github.com/Kong/shared-actions/blob/main/.github/workflows/ci.yaml#L43-L47

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
+name: CI
+
+on:
+  pull_request:
+    branches:
+      - main
```
Comment (saisatishkarra) [inline .github/workflows/ci.yml:6]: Include push to main

**Comment by saisatishkarra** [comment]
TODO / Missing things:
- How are the conventional commits being enforced on local machine? Lefthook / husky git hooks for pre-commit can be integrated with commitzen and conventional commits to ensure the commit messages structure. Check: https://github.com/Kong/shared-actions/blob/main/lefthook.yml
- Update README on how to run and perform git commit with the installed pre-commit hooks
- Update the `commitlint.config.js` to check for the package names to be used within commit messages during linting.  Ex: https://github.com/Kong/shared-actions/commit/ab62f80e83ff8e11e7c1106ace62c5842952a71f is the sample commit message
-

**Review by saisatishkarra — CHANGES_REQUESTED** [review]

File: .github/workflows/auto-tagger.yml
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply .github/workflows/auto-tagger.yml:1]: - It makes sense to remove action-tagger workflow now as we are not releasing the PSA as a whole and are decoupling the releases and have decided to use tags like project-example@x.x.x and not shorthands like v1 or latest.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (pankajmouriyakong) [reply .github/workflows/ci.yml:6]: done

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
+name: Releases
+on:
+  push:
+    branches:
+      - main
+  workflow_dispatch: {}
+
+jobs:
+  release:
+    if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' && github.repository_owner == 'pankajmouriyakong'}}
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply .github/workflows/release.yml:10]: This is not needed here. I will remove this. I was supposed to remove this.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    name: Release
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Generate privileged push token to bypass branch protection settings
+        id: github-app-token
+        uses: tibdex/github-app-token@b62528385c34dbc9f38e5f4225ac829252d1ea92 # v1.8.0
+        with:
+          app_id: ${{ secrets.APP_ID }}
+          private_key: ${{ secrets.PRIVATE_KEY }}
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply .github/workflows/release.yml:20]: This is a forked repo and no these are not available as org wide secrets. 

When raising the PR in PSA, a GitHub App will be created and a private key will be provisioned if there is no existing app already provisioned for such use cases. 

PSA repo secrets should be good enough for storing the private key and app ID as `release.yml` will run within PSA only. I will add a ToDo note for creating an app within the Kong org.

File: .github/workflows/release.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply .github/workflows/release.yml:10]: done

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+      # This step will create a tag and also release
+      - name: Release new versions
+        env:
+          GH_TOKEN: ${{ steps.github-app-token.outputs.token }}
+        shell: bash
+        run: |
+          git config --local user.name "${{ vars.GH_APP__KONG__BOT_USERNAME }}"
+          git config --local user.email "${{ vars.GH_APP__KONG__BOT_EMAIL }}"
+          pnpm version:ci
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply .github/workflows/release.yml:59]: added

File: package.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
+{
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply package.json:1]: done

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        shell: bash
+        run: pnspm i
+
+      - name: Release new versions
+        shell: bash
+        run: |
+          git checkout ${{ github.head_ref }}
+          git config --local user.name "${{ vars.GH_APP__KONG__BOT_USERNAME }}"
+          git config --local user.email "${{ vars.GH_APP__KONG__BOT_EMAIL }}"
+          spnpm version:dry-run
```
Comment (saisatishkarra) [inline .github/workflows/ci.yml:52]:
```suggestion
pnpm version:dry-run
```

Comment (pankajmouriyakong) [reply .github/workflows/ci.yml]: hold, I am testing failure slack notifications @saisatishkarra

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        with:
+          path: '**/node_modules'
+          key: pnpm-20-${{ runner.os }}-${{ hashFiles('**/pnpm-lock.yaml') }}
+
+      - name: Build & Test action
+        if: ${{ steps.dependency-cache.outputs.cache-hit != 'true' && steps.filter.outputs.scan-docker-image == 'true'}}
+        shell: bash
+        run: |
+          git checkout ${{ github.head_ref }}
+          pnpm install  --frozen-lockfile
```
Comment (saisatishkarra) [inline .github/workflows/ci.yml:61]:
```suggestion
          pnpm install --frozen-lockfile
```

File: security-actions/login-aws-cloud/action.yml
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply security-actions/login-aws-cloud/action.yml:1]: remove

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        with:
+          path: '**/node_modules'
+          key: pnpm-20-${{ runner.os }}-${{ hashFiles('**/pnpm-lock.yaml') }}
+
+      - name: Install Dependencies
+        if: ${{ steps.dependency-cache.outputs.cache-hit != 'true' }}
+        shell: bash
+        run: pnpm i
+
+      - name: Release new versions
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply .github/workflows/ci.yml]: changes made as requested

File: .github/workflows/ci.yml (hunk 1 above)
Comment (pankajmouriyakong) [reply .github/workflows/ci.yml]: changed updated

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+      - name: Install PNPM
+        run: |
+          npm install -g pnpm
+          pnpm --version
+          if [ -n "${{ github.head_ref }}" ]; then
+            git checkout ${{ github.head_ref }}
+          fi
+
+      - name: Dependency Cache
```
Comment (saisatishkarra) [inline .github/workflows/ci.yml:66]: Caching pnpm `node_modules (optimizes build speed but might not always work and not resilient to changing lock file)`.

Instead, only cache the `~/.pnpm-store` (global cache) and `pnpm install --frozen-lockfile` each time to reconstruct node_modules ( takes a bit time but efficient and we have very few dependencies at the moment to rebuild) which allows for accomadating minor changes for lock files during PR updates and reduces the cache invalidation frequently

For most use cases with PNPM:
- Refer: https://github.com/orgs/pnpm/discussions/6374 
- Cache only ~/.pnpm-store: This is the default and preferred approach for CI pipelines.

Bottomline: Remove node_modules caching for now.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    runs-on: ubuntu-latest
+    if: ${{ needs.changes.outputs.packages != '[]' }}
+    outputs:
+      nodeVersion: ${{ steps.node-version.outputs.nodeVersion }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Setup Node
```
Comment (saisatishkarra) [inline .github/workflows/ci.yml:47]: Refer pnpm global package caching at https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#caching-packages-data 

setup-pnpm and setup-node with cache input `pnpm` and `cache-depdency-path` should be used which by default caches `~/.pnpm-store` instead of `node_modules`

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        package: ${{ fromJSON(needs.changes.outputs.packages) }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Setup Node
```
Comment (saisatishkarra) [inline .github/workflows/ci.yml:96]: To automatically restore global cache in the `setup-dependencies` job, Refer pnpm global package caching at https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#caching-packages-data

```
- uses: pnpm/action-setup@v2
  with:
    version: <REPLACE>
- uses: actions/setup-node@v4
  with:
    node-version: '<REPLACE>'
    cache: 'pnpm' # automatically restores pnpm global cache but not node_modules from the previous job if available
```

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        uses: actions/setup-node@v4
+        with:
+          node-version: 20
+
+      - name: Install PNPM
+        run: |
+          npm install -g pnpm
+          pnpm --version
+
+      - name: Dependency Cache
```
Comment (saisatishkarra) [inline .github/workflows/ci.yml:106]: Remove this

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        uses: actions/cache@v3
+        with:
+          path: '**/node_modules'
+          key: pnpm-${{ needs.setup-dependencies.outputs.nodeVersion }}-${{ runner.os }}-${{ hashFiles('**/pnpm-lock.yaml') }}
+
+      - name: Restore Dependencies
+        if: ${{ steps.dependency-cache.outputs.cache-hit != 'true' }}
```
Comment (saisatishkarra) [inline .github/workflows/ci.yml:113]: always reconstruct node_modules using `pnpm install --frozen-lockfile`. Pnpm and setup node will automatically optimize this.

File: .github/workflows/release.yml [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+          app_id: ${{ secrets.APP_ID }}
+          private_key: ${{ secrets.PRIVATE_KEY }}
+
+      - name: Checkout code
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+          token: ${{ steps.github-app-token.outputs.token }}
+
+      - name: Set up Node.js
```
Comment (saisatishkarra) [inline .github/workflows/release.yml:35]: To automatically restore global cache in the `setup-dependencies` job, Refer pnpm global package caching at https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#caching-packages-data

```
- uses: pnpm/action-setup@v2
  with:
    version: <REPLACE>
- uses: actions/setup-node@v4
  with:
    node-version: '<REPLACE>'
    cache: 'pnpm' # automatically restores pnpm global cache but not node_modules from the previous job if available
```

File: .github/workflows/release.yml [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+        with:
+          node-version: '20'
+
+      - name: Install PNPM
+        shell: bash
+        run: |
+          npm i -g pnpm
+          pnpm --version
+
+      - name: Dependency Cache
```
Comment (saisatishkarra) [inline .github/workflows/release.yml:46]: Remove this

File: .github/workflows/release.yml [hunk 17]
```diff
@@ -0,0 +1,123 @@
+name: Publish Release
+on:
+  push:
+    branches:
+      - main
+  pull_request:
```
Comment (saisatishkarra) [inline .github/workflows/release.yml:6]: Remove pull_request event for release. Why is this needed?

**Review by saisatishkarra — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+      - name: Build & Test action
+        if: ${{ steps.dependency-cache.outputs.cache-hit != 'true' && steps.filter.outputs.scan-docker-image == 'true'}}
+        shell: bash
+        run: |
+          git checkout ${{ github.head_ref }}
+          pnpm install --frozen-lockfile
+          pnpm lint
+          pnpm run package
+
+
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply .github/workflows/ci.yml:97]: done

File: .github/workflows/release.yml (hunk 16 above)
Comment (pankajmouriyakong) [reply .github/workflows/release.yml:50]: done

File: README.md [hunk 19]
```diff
@@ -3,14 +3,106 @@ Shared actions available to both public and private repositories
...
+
+For example:
   
-  
\ No newline at end of file
+```yaml
+- uses:  Kong/public-shared-actions/code-build-actions/build-js-sdk@v1.6.0
+```
+
+# Setting Up Lefthook for Commit Message Verification
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply README.md:16]: Official doc link mentioned in the step  `### For Other Platforms`

File: README.md [hunk 20]
```diff
@@ -3,14 +3,106 @@ Shared actions available to both public and private repositories
...
+```bash
+snap install --classic lefthook
+```
+
+### For Other Platforms
+Refer to the official Lefthook [installation guide](https://github.com/evilmartians/lefthook/blob/master/docs/install.md) for platform-specific instructions.
+
+---
+
+## Step 2: Initialize Lefthook
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply README.md:39]: agree on this, made the required change.

File: commitlint.config.js [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+    // see: https://github.com/conventional-changelog/commitlint/blob/master/docs/reference-rules.md
+    'scope-empty': [2, 'never'],
+    'scope-enum': async (ctx) => [
+      2,
+      'always',
+      [...(await getPackages(ctx)), 'release', 'deps', 'ci']
+    ]
+  },
+  helpUrl: 'https://github.com/Kong/public-shared-actions',
+  prompt: {
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply commitlint.config.js:17]: This is optional, users can run the standard `git commit` command to commit their changes. 

If someone wants to follow the interactive prompt way, they need to run `pnpm commit` instead of `git commit`. I have updated the README with this.

File: .github/workflows/release.yml [hunk 22]
```diff
@@ -0,0 +1,111 @@
+name: Publish Release
+on:
+  push:
```
[the remark this answers is not in the document]
Comment (pankajmouriyakong) [reply .github/workflows/release.yml]: Change updated,  testing it progress
